logstash/class-logger.php

```diff
@@ -0,0 +1,426 @@
+			return; // Failed validation.
+		}
+
+		// Adjust `feature` and JSON-encode pretty-print data.
+		$data['feature']        = 'a8c_vip_' . $data['feature'];
+
+		// Adjust `extra` and JSON-encode data that gets sent in the API request.
+		$data['extra'] = isset( $data['extra'] ) ? wp_json_encode( $data['extra'], JSON_PRETTY_PRINT ) : '';
+
+		// Adds a new log entry.
+		static::$entries[] = $data;
+
+		// Sends data to logstash on shutdown.
+		if ( ! has_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] ) ) {
+			add_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] );
+		}
+	}
+
+	/**
+	 * Sends data to logstash via REST API on shutdown, viewable in Kibana.
+	 *
+	 * @since 2020-01-10
+	 */
+	public static function process_entries_on_shutdown() : void {
+		if ( static::$processed_entries ) {
+			return; // Already done.
+		}
+		static::$processed_entries = true;
+		$endpoint = 'https://public-api.wordpress.com/rest/v1.1/logstash/bulk';
+
+		$entry_chunks = array_chunk( static::$entries, static::BULK_ENTRIES_COUNT );
+		// Process all entries.
+		foreach ( $entry_chunks as $entries ) {
+			if ( ! defined( 'VIP_GO_ENV' ) || ! VIP_GO_ENV ) {
+				static::maybe_wp_debug_log_entries( $entries );
+				continue; // Bypassing REST API below in this case.
+			}
+			$json_data     = wp_json_encode( $entries );
+			// Send to logstash via REST API endpoint with payload containing log entry details.
+			$_wp_remote_response         = wp_remote_post(
+				$endpoint,
+				[
+					'timeout'     => 2,
+					'redirection' => 0,
+					'blocking'    => false,
+					'body'        => [
+						'params' => $json_data,
+					],
+				]
+			);
+			$_wp_remote_response_code    = wp_remote_retrieve_response_code( $_wp_remote_response );
+			$_wp_remote_response_message = wp_remote_retrieve_response_message( $_wp_remote_response );
+
+			if ( 200 !== $_wp_remote_response_code ) {
+				static::maybe_wp_debug_log_entries( $entries );
+				trigger_error( 'Unable to ' . esc_html( __METHOD__ ) . '(). Response from <' . esc_html( $endpoint ) . '> was [' . esc_html( $_wp_remote_response_code ) . ']: ' . esc_html( $_wp_remote_response_message ) . '.', E_USER_WARNING );
+			}
+		}
+	}
+
+	/**
+	 * Sends data to `WP_DEBUG_LOG` when applicable.
+	 *
+	 * @since 2020-01-10
+	 *
+	 * @param array $entries.
+	 */
+	public static function maybe_wp_debug_log_entries( array $entries ) : void {
+		if ( ! apply_filters( 'enable_wp_debug_mode_checks', true ) ) {
+			return; // Not applicable.
+		} elseif ( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) {
+			return; // Not applicable.
+		} elseif ( ! defined( 'WP_DEBUG_LOG' ) || ! WP_DEBUG_LOG ) {
+			return; // Not applicable.
+		}
+
+		foreach ( $entries as $entry ) {
+			static::wp_debug_log( $entry );
+		}
+	}
+
+	/**
+	 * Save message to log file
+	 *
+	 * @since 2020-01-10
+	 *
+	 * @param array $entry Data.
+	 */
+	public static function wp_debug_log( array $entry ) : void {
```

Comment:
We need extra protections here against this running on the Go infrastructure. Since it's a public function, any code can call it, which will attempt (and hopefully fail) to write to `WP_CONTENT_DIR . '/debug.log'`.

I don't think there is a valid case for this to run on Go (only on sandboxes or locally), so we should check the constants to ensure it's only allowed in those spots.

Reply:
Nice catch! Added in 51df58d0e420dcc8590d687a967f5f12fcb4be04